src/celeritas/neutron/interactor/detail/CascadeCollider.hh

```diff
@@ -0,0 +1,235 @@
+};
+
+//---------------------------------------------------------------------------//
+// Inline DEFINITIONS
+//---------------------------------------------------------------------------//
+/*!
+ * Construct from share data and colliding nucleons in the lab frame.
+ */
+CELER_FUNCTION
+CascadeCollider::CascadeCollider(NeutronInelasticRef const& shared,
+                                 CascadeParticle const& bullet,
+                                 CascadeParticle const& target)
+    : shared_(shared)
+    , bullet_(bullet)
+    , target_(target)
+    , ch_id_(ChannelId{
+          static_cast<size_type>((bullet_.type == target_.type) ? 0 : 1)})
+    , sample_phi_(0, 2 * constants::pi)
+{
+    // Initialize the boost velocity and momentum in the center of mass frame
+    FourVector sum_four_vec = bullet_.four_vec + target_.four_vec;
+    cm_velocity_ = boost_vector(sum_four_vec);
+    cm_p_ = this->calc_cm_p(sum_four_vec);
+
+    // Calculate the kinetic energy in the target rest frame
+    FourVector bullet_p = bullet_.four_vec;
+    boost(-boost_vector(target_.four_vec), &bullet_p);
+    kin_energy_ = bullet_p.energy - norm(bullet_p);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Sample the final state of outgoing particles from the two-body intra-nucleus
+ * nucleon-nucleon collision in the center of mass (c.m.) frame and return
+ * them after converting momentum to the lab frame. This performs the same
+ * sampling routine as in Geant4's G4ElementaryParticleCollider, mainly
+ * implemented in collide and generateSCMfinalState methods. The cos\theta
+ * distribution in c.m. is inversely sampled using the tabulated cumulative
+ * distribution function (c.d.f) data in the kinetic energy and cos\theta bins
+ * which are implemented in G4CascadeFinalStateAlgorithm::GenerateTwoBody and
+ * G4NumIntTwoBodyAngDst::GetCosTheta methods.
+ */
+template<class Engine>
+CELER_FUNCTION auto CascadeCollider::operator()(Engine& rng) -> FinalState
+{
+    // Sample cos\theta of outgoing particles in the center of mass frame
+    real_type cdf = generate_canonical(rng);
+    TwodGridData cdf_grid = shared_.angular_cdf[ch_id_];
+    Grid energy_grid(cdf_grid.x, shared_.reals);
+    real_type cos_theta{0};
+
+    if (kin_energy_ < energy_grid.back())
+    {
+        // Find cos\theta from tabulated angular data for a given c.d.f.
+        Grid cos_grid(cdf_grid.y, shared_.reals);
+        TwodGridCalculator calc_cdf(cdf_grid, shared_.reals);
+
+        size_type idx = cos_grid.size() - 2;
+        real_type cdf_upper = 0;
+        real_type cdf_lower = 1;
+
+        do
+        {
+            cdf_upper = cdf_lower;
+            cdf_lower = calc_cdf({kin_energy_, cos_grid[idx]});
+        } while (cdf_lower > cdf && idx-- > 0);
+
+        real_type frac = (cdf - cdf_lower) / (cdf_upper - cdf_lower);
+        cos_theta = fma(frac, cos_grid[idx + 1] - cos_grid[idx], cos_grid[idx]);
```

Comment:
Great!  I originally tried an array of GenericGridData [energy bin], so that we can swap x an y after find a corresponding cos_grid_data for each energy bin, then use the GenericGridCalculator to sample cos(theta). I believe that two approaches are basically equivalent, but using Two2GridData is much more compact and cleaner.  If we can encapsulate the inverse sampling inside Twod(Sub)gridCalculator, it would be even great!  Thanks for your help.